### PR TITLE
feat(client): implement rejectOnNotFound 

### LINKED ISSUE #5047: Add rejectOnNotFound to findOne/findFirst
## Problem
Instead of returning a nullable result when calling `model.findOne`, accept a new argument (`rejectOnEmpty`) that makes the returning type non-nullable and throws an error when the row was not found

## Suggested solution

```ts
const person = await prisma.person.findOne({ where: { id } })

if (person) {
  // do stuff
}
```

Will be:

```ts
const person = await prisma.person.findOne({ where: { id }, rejectOnEmpty: true })

// do stuff
```

`rejectOnEmpty` will accept either boolean or an `Error` so the thrown error can be customized:

```ts
const person = await prisma.person.findOne({ 
  where: { id }, 
  rejectOnEmpty: new MyCustomNotFoundError(`person with id=${id} was not found`)
})
```

## Additional context

This API is stolen from Sequelize

### PR DESCRIPTION
**Pull request by williamluke4**
Closes https://github.com/prisma/prisma-client-js/issues/708
Closes https://github.com/prisma/prisma/issues/5047

### CODE REVIEW

File: src/packages/client/src/__tests__/integration/happy/exhaustive-schema/__snapshots__/test.ts.snap [hunk 1]
```diff
@@ -3885,6 +3897,10 @@ export namespace Prisma {
      * Choose, which related nodes to fetch as well.
     **/
     include?: MInclude | null
+    /**
+     * Choose, which related nodes to fetch as well.
```
Comment (williamluke4) [inline src/packages/client/src/__tests__/integration/happy/exhaustive-schema/__snapshots__/test.ts.snap:3901]: Fix

File: src/packages/client/src/__tests__/integration/happy/exhaustive-schema/__snapshots__/test.ts.snap [hunk 2]
```diff
@@ -1406,6 +1406,10 @@ export namespace Prisma {
   export type ErrorFormat = 'pretty' | 'colorless' | 'minimal'
 
   export interface PrismaClientOptions {
+    /**
+     * Will throw an Error if findUnique returns null
```
Comment (williamluke4) [inline src/packages/client/src/__tests__/integration/happy/exhaustive-schema/__snapshots__/test.ts.snap:1410]: Improve the JSDoc

File: src/packages/client/src/__tests__/types/rejectOnEmpty/test.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
+import { PrismaClient } from '@prisma/client'
+
+// tslint:disable
+const cases = {
```
Comment (williamluke4) [inline src/packages/client/src/__tests__/types/rejectOnEmpty/test.ts:4]: Remove this

File: src/packages/client/src/generation/TSClient/Args.ts [hunk 4]
```diff
@@ -69,7 +69,21 @@ export class ArgsType implements Generatable {
...
+        isRequired: false,
+        isNullable: true,
+        inputTypes: [
+          {
+            type: 'boolean | Error',
+            location: 'scalar',
+            isList: false,
+          },
+        ],
+        comment: `Choose, which related nodes to fetch as well.`,
```
Comment (williamluke4) [inline src/packages/client/src/generation/TSClient/Args.ts:84]: Fix

File: src/packages/client/src/runtime/getPrismaClient.ts [hunk 5]
```diff
@@ -1032,6 +1037,22 @@ new PrismaClient({
       const { isList } = field.outputType
       const typeName = getOutputTypeName(field.outputType.type)
 
+      let rejectOnEmpty:  boolean | Error;
+      if(args && typeof args === 'object' && 'rejectOnEmpty' in args && args.rejectOnEmpty !== undefined) {
+        console.log('Args');
+        rejectOnEmpty = args['rejectOnEmpty']
+        delete args['rejectOnEmpty'];
+      } else if(this._rejectOnEmpty && typeof this._rejectOnEmpty === 'object' && typeName in this._rejectOnEmpty){
+        console.log('Object');
```
Comment (williamluke4) [inline src/packages/client/src/runtime/getPrismaClient.ts:1046]: Remove logs

File: src/packages/client/src/runtime/getPrismaClient.ts [hunk 6]
```diff
@@ -1570,6 +1594,13 @@ export class PrismaClientFetcher {
       if (process.env.PRISMA_CLIENT_GET_TIME) {
         return { data: unpackResult, elapsed }
       }
+      if(clientMethod.includes('findUnique') && rejectOnEmpty){
```
Comment (williamluke4) [inline src/packages/client/src/runtime/getPrismaClient.ts:1597]: @timsuchanek is there a better way to do this?

File: src/packages/client/src/runtime/getPrismaClient.ts [hunk 7]
```diff
@@ -1570,6 +1594,13 @@ export class PrismaClientFetcher {
...
+        if(typeof rejectOnEmpty === 'boolean'){
```
Comment (williamluke4) [inline src/packages/client/src/runtime/getPrismaClient.ts:1598]: And true?

File: src/packages/client/src/runtime/utils/validatePrismaClientOptions.ts [hunk 8]
```diff
@@ -161,6 +162,17 @@ It should have this form: { url: "CONNECTION_STRING" }`,
...
   },
+  rejectOnEmpty: (value => {
+    if (!value) {
+      return
+    }
+    if(isError(value) || typeof value === 'boolean' || typeof value === 'object'){
+      return value
+    }
+    throw new PrismaClientConstructorValidationError(
+      `Invalid rejectOnEmpty expected a boolean/Error/{[modelName: Error | boolean]} but received ${JSON.stringify(value)}`,
```
Comment (williamluke4) [inline src/packages/client/src/runtime/utils/validatePrismaClientOptions.ts:173]: Create better error messaage

**Review by williamluke4 — COMMENTED** [review]
Needs a bit more work before it's ready to go

File: src/packages/client/fixtures/blog/prisma/schema.prisma [hunk 9]
```diff
@@ -16,7 +16,6 @@ model User {
   age        Int?
   posts      Post[]
   likes      Like[]
```
Comment (timsuchanek) [inline src/packages/client/fixtures/blog/prisma/schema.prisma:18]: Please keep this

File: src/packages/client/src/__tests__/integration/happy/rejectNotFound/test.ts [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+      customError: new Error('FindFirst Custom Error'),
+      thunk: () => new Error('FindFirst Thunk'),
+      true: true,
+      false: false,
+      undefined: undefined,
+    },
+  },
+}
+
+for (const constructorKey of Object.keys(cases.constructor)) {
```
Comment (timsuchanek) [inline src/packages/client/src/__tests__/integration/happy/rejectNotFound/test.ts:49]: Noice

File: src/packages/client/src/generation/TSClient/PrismaClient.ts [hunk 11]
```diff
@@ -73,7 +73,7 @@ export class PrismaClient<
 
   ${indent(this.jsDoc, TAB_SIZE)}
 
-  constructor(optionsArg ?: T);
+  constructor(optionsArg ?: Prisma.Subset<T, Prisma.PrismaClientOptions>);
```
Comment (timsuchanek) [inline src/packages/client/src/generation/TSClient/PrismaClient.ts:76]: Thanks

File: src/packages/client/src/generation/TSClient/PrismaClient.ts [hunk 12]
```diff
@@ -158,9 +158,29 @@ get ${methodName}(): Prisma.${m.model}Delegate;`
   public toTS(): string {
     return `${new Datasources(this.internalDatasources).toTS()}
 
+export type RejectOnNotFound = boolean | Error | ((error: Error) => Error)
+type LooseRecord<K extends string | number | symbol, T> = { [P in K]?: T; }
```
Comment (timsuchanek) [inline src/packages/client/src/generation/TSClient/PrismaClient.ts:162]: Can you elaborate why we need a `LooseRecord`?

**Review by timsuchanek — COMMENTED** [review]
Good job! This already looks very good :+1:

File: src/packages/client/src/generation/TSClient/PrismaClient.ts (hunk 12 above)
Comment (williamluke4) [reply src/packages/client/src/generation/TSClient/PrismaClient.ts]: Not needed so i've removed it

**Pull request merged**